File: OLS_likelihood.R
# Simulated data
set.seed(12345)
n <- 1000
k <-2
X <- cbind(1, matrix(rnorm(n * k), n, k))
true_coeff <- rnorm(k,1,1)
y <- X %*% true_coeffs + rnorm(n)

#Likelihood function
ols_likelihood <- function(params,data) {
  #parameters
  beta <- params
  #Y and X definition
  X <- data$X
  y <- data$y
  ###########
  predicted <- X %*% beta
  residuals <- y - predicted
  #Calculated log-likelihood to normal distribution (for residual terms)
  log_likelihood <- sum(dnorm(residuals, mean = 0, sd = 1, log = TRUE))
  return(log_likelihood)
}

#initial point
initial_params <- rep(1,ncol(X))
#Data for likelihood function
data <- list(X = X, y = y)
#Solving
result <- optim(initial_params, ols_likelihood, data = data, method = "BFGS")
print(result$par)

File: OLS_likelihood.py
import numpy as np
from scipy.optimize import minimize

# Simulated data
np.random.seed(12345)
n = 1000
k = 2
X = np.hstack((np.ones((n, 1)), np.random.randn(n, k)))
true_coeffs = np.random.normal(1, 1, k)
y = X.dot(true_coeffs) + np.random.randn(n)

# Likelihood function
def ols_likelihood(params, data):
    beta = params
    X = data['X']
    y = data['y']
    
    predicted = X.dot(beta)
    residuals = y - predicted
    
    log_likelihood = np.sum(np.log(np.exp(-0.5 * ((residuals - 0) / 1)**2) / np.sqrt(2 * np.pi)))
    return -log_likelihood

# Initial point
initial_params = np.ones(X.shape[1])
# Data for likelihood function
data = {'X': X, 'y': y}

# Solving
result = minimize(ols_likelihood, initial_params, args=(data,), method='BFGS')
print(result.x)
